rfcs/20220523-depthwise-post-op-extension/README.md

```diff
@@ -0,0 +1,64 @@
+# RFC: Depthwise post-op extension
+
+## Introduction
+oneDNN provides a performance optimization for a chain of convolutions, when 1x1 convolution is followed by 3x3 depthwise convolution with stride of 1 or 2.
+This optimization is provided through an attribute post-op API.
+To keep usability in place, the decision in the past was to hardcode depthwise convolution settings into a post-op name.
+Thus, `k3s1p1` or `k3s2p1` options are only allowed.
+This optimization targeted only MobileNet topologies popular back then.
+Now there are models that have same pattern of 1x1 + depthwise, but they might have different depthwise convolution settings.
+E.g., they could have kernel size of 3, 5, or 7; stride size of 1, or 2; (left) padding size of 0, or 1.
+In addition, it may happen that certain convolutions have right padding greater than left, especially when padding size is 0.
+Extended API will allow to fuse those convolutions and may provide 5%-15% speed up on a model level as measured by TensorFlow benchmarks.
+
+## Proposal
+Support more shapes that appear for depthwise convolutions through a more general API for depthwise post-op in oneDNN.
+Instead of hardcoding parameters in the name, add them as an explicit argument for a call.
+If parameters form a valid convolution setup, at least the implementation that calls two optimized implementations should be invoked.
+Performance assumptions hold for extended API - it provides better, or on-par performance compared to unfused implementation.
+Data type coverage remains the same as it was before.
+The API targets the oneDNN v2.6 Update release to intercept the next TensorFlow v2.10 release.
+The next oneDNN v3.0 Major release will have previous version of API removed since new one covers previously supported cases.
+
+## API
+```c
+/// dnnl.h
+
+dnnl_status_t DNNL_API dnnl_post_ops_append_dw(dnnl_post_ops_t post_ops,
+        dnnl_data_type_t weights_data_type, dnnl_data_type_t bias_data_type,
+        dnnl_data_type_t dst_data_type, dnnl_dim_t kernel_size,
+        dnnl_dim_t stride_size, dnnl_dim_t padding_l_size, dnnl_dim_t count,
+        int mask, const float *scales);
```

Comment:
If our implementation permits, I would advocate to not pass scales here, but accept only runtime scales.

Reply:
I wish it does but it doesn't. Do you think I need to spend some time on enabling it before publishing new API?
It will require benchdnn CLI extending and all convolutions that support a single run (not through fused_ref) modifications.

Reply:
I guess the implementation will have to be adjusted to support runtime scales anyway if we decide to move away from creation time scales. Unless we want to keep creation time scales for this case only we should change this API.

Reply:
> I wish it does but it doesn't. Do you think I need to spend some time on enabling it before publishing new API?

The question is mostly about timeline I guess. The change to runtime scales will have to happen by v3 at least.
So if the targeted users for this feature :
- have a tight timeline that we cannot meet if we move to runtime scales first
- and are ok to use static scales first, and migrate to runtime scales with v3 migration,
I guess we are fine to proceed with static scales. Though I would really encourage to go with runtime scales to reduce disruption upon migration to v3 (the more we disrupt, the longer it takes for users to migrate and get the optimizations we will put there).

Reply:
I'll ask about timelines, but the primary target was not an int8 flow, so scaling shouldn't be a deal breaker.

> Though I would really encourage to go with runtime scales to reduce disruption upon migration to v3...
So, none of CPU implementations support runtime scales, so at this moment it means to enable all convolutions first, and only then dw post-op. So I'd drop the support as proposed and we update API later so that users may get benefits of the fusion.

> Could you clarify which extensions are needed there?
I initially thought it will require modifications but now I see it's not, only internal implementation to pass these scales to the library.

Reply:
> So I'd drop the support as proposed and we update API later so that users may get benefits of the fusion.

Could you clarify ? From the discussion, if f32 is the primary target, I would think that it does not matter which version we use. So why not just go with the definitive one directly ?

For example, we could first have in C API (reserved argument for mask):
```c
void append_dw(memory::data_type weights_data_type,
            memory::data_type bias_data_type, memory::data_type dst_data_type,
            memory::dim kernel_size, memory::dim stride_size,
            memory::dim padding_l_size, int reserved);
```

And in C++ API (signature with no int8 specific argument):
```c++
void append_dw(memory::data_type weights_data_type,
            memory::data_type bias_data_type, memory::data_type dst_data_type,
            memory::dim kernel_size, memory::dim stride_size,
            memory::dim padding_l_size);

```

And once we add support for int8 scales to convolutions/dw post op, we would modify the C and C++ API accordingly (with proper argument naming and proper overload respectively)

C API (proper naming of reserved argument)
```c
void append_dw(memory::data_type weights_data_type,
            memory::data_type bias_data_type, memory::data_type dst_data_type,
            memory::dim kernel_size, memory::dim stride_size,
            memory::dim padding_l_size, int mask);
```

C++ API (add an overload for int8 extra argument)
```c++
void append_dw(memory::data_type weights_data_type,
            memory::data_type bias_data_type, memory::data_type dst_data_type,
            memory::dim kernel_size, memory::dim stride_size,
            memory::dim padding_l_size, int mask);
```

This should allow to both not break code of users that will adopt this new function, and allow later introduction with proper signatures for int8 handling. What do you think?

Reply:
Discussed offline. The feature is under time pressure, but requires some significant changes on validation side. Thanks for the suggestion, we will update API with runtime scales enabling.